In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns



In [2]:
import os

# Carpeta de notebooks
ruta_proyecto = os.path.abspath("..")
print("Ruta proyecto:", ruta_proyecto)

# Listar todo lo que tiene
print("Contenido de proyecto_final:")
print(os.listdir(ruta_proyecto))


Ruta proyecto: c:\Users\ffmma\OneDrive\Desktop\proyecto_final
Contenido de proyecto_final:
['.git', '.gitignore', 'data', 'docs', 'notebooks', 'README.md', 'src']


In [3]:
import os

ruta_data = os.path.join("..", "data")
print("Contenido de data/:")
print(os.listdir(ruta_data))


Contenido de data/:
['data_processed', 'data_raw']


In [4]:
# Carpeta donde están los archivos
ruta_data_raw = os.path.join("..", "data", "data_raw")
ruta_data_raw

'..\\data\\data_raw'

In [5]:
archivo1 = os.path.join(ruta_data_raw, "telecom_clients_us.csv")

df1 = pd.read_csv(archivo1)

df1


user_id tariff_plan  date_start
0     166713           A  2019-08-15
1     166901           A  2019-08-23
2     168527           A  2019-10-29
3     167097           A  2019-09-01
4     168193           A  2019-10-16
..       ...         ...         ...
727   166554           B  2019-08-08
728   166911           B  2019-08-23
729   167012           B  2019-08-28
730   166867           B  2019-08-22
731   166565           B  2019-08-08

[732 rows x 3 columns]

In [6]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 732 entries, 0 to 731
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      732 non-null    int64 
 1   tariff_plan  732 non-null    object
 2   date_start   732 non-null    object
dtypes: int64(1), object(2)
memory usage: 17.3+ KB


In [7]:
#cambio de tipos de la columna date_start de objetct a datetime 
df1["date_start"]=pd.to_datetime(df1["date_start"], errors="coerce")

df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 732 entries, 0 to 731
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   user_id      732 non-null    int64         
 1   tariff_plan  732 non-null    object        
 2   date_start   732 non-null    datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(1)
memory usage: 17.3+ KB


In [8]:
#chequo de valores nulos
df1.isnull().sum()
#ninguna columna tiene valores nulos 


user_id        0
tariff_plan    0
date_start     0
dtype: int64

In [9]:
df1.head()

user_id tariff_plan date_start
0   166713           A 2019-08-15
1   166901           A 2019-08-23
2   168527           A 2019-10-29
3   167097           A 2019-09-01
4   168193           A 2019-10-16

In [10]:
archivo2 = os.path.join(ruta_data_raw, "telecom_dataset_us.csv")

df2 = pd.read_csv(archivo2)

df2.head()

user_id                       date direction internal  operator_id  \
0   166377  2019-08-04 00:00:00+03:00        in    False          NaN   
1   166377  2019-08-05 00:00:00+03:00       out     True     880022.0   
2   166377  2019-08-05 00:00:00+03:00       out     True     880020.0   
3   166377  2019-08-05 00:00:00+03:00       out     True     880020.0   
4   166377  2019-08-05 00:00:00+03:00       out    False     880022.0   

   is_missed_call  calls_count  call_duration  total_call_duration  
0            True            2              0                    4  
1            True            3              0                    5  
2            True            1              0                    1  
3           False            1             10                   18  
4            True            3              0                   25

In [11]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53902 entries, 0 to 53901
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   user_id              53902 non-null  int64  
 1   date                 53902 non-null  object 
 2   direction            53902 non-null  object 
 3   internal             53785 non-null  object 
 4   operator_id          45730 non-null  float64
 5   is_missed_call       53902 non-null  bool   
 6   calls_count          53902 non-null  int64  
 7   call_duration        53902 non-null  int64  
 8   total_call_duration  53902 non-null  int64  
dtypes: bool(1), float64(1), int64(4), object(3)
memory usage: 3.3+ MB


In [12]:
df2.sample(10)

user_id                       date direction internal  operator_id  \
44539   168062  2019-11-01 00:00:00+03:00       out    False     947612.0   
4664    166511  2019-10-01 00:00:00+03:00        in    False     891410.0   
15162   166884  2019-10-28 00:00:00+03:00        in    False          NaN   
30158   167285  2019-10-16 00:00:00+03:00       out    False     908640.0   
40537   167827  2019-10-12 00:00:00+03:00        in    False          NaN   
39181   167716  2019-10-11 00:00:00+03:00       out    False     922142.0   
23711   167109  2019-09-16 00:00:00+03:00       out     True     906294.0   
31229   167412  2019-11-16 00:00:00+03:00        in     True          NaN   
495     166377  2019-11-13 00:00:00+03:00       out    False     880028.0   
40922   167828  2019-10-31 00:00:00+03:00        in    False     940438.0   

       is_missed_call  calls_count  call_duration  total_call_duration  
44539            True            1              0                    8  
4664            False           10            611                  744  
15162            True            1              0                    2  
30158            True          366              0                 4915  
40537            True            1              0                    1  
39181           False            7            783                  957  
23711            True            1              0                    0  
31229            True            1              0                    2  
495              True            7              0                  180  
40922           False            4            563                  583

In [13]:
#cambio a datetime de la columna operator_id
df2["date"]= pd.to_datetime(df2["date"],errors = "coerce")
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53902 entries, 0 to 53901
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype                    
---  ------               --------------  -----                    
 0   user_id              53902 non-null  int64                    
 1   date                 53902 non-null  datetime64[ns, UTC+03:00]
 2   direction            53902 non-null  object                   
 3   internal             53785 non-null  object                   
 4   operator_id          45730 non-null  float64                  
 5   is_missed_call       53902 non-null  bool                     
 6   calls_count          53902 non-null  int64                    
 7   call_duration        53902 non-null  int64                    
 8   total_call_duration  53902 non-null  int64                    
dtypes: bool(1), datetime64[ns, UTC+03:00](1), float64(1), int64(4), object(2)
memory usage: 3.3+ MB


In [14]:
#extracción de fecha y horas de la columna date
df2["date"] = df2["date"].dt.date

df2


user_id        date direction internal  operator_id  is_missed_call  \
0       166377  2019-08-04        in    False          NaN            True   
1       166377  2019-08-05       out     True     880022.0            True   
2       166377  2019-08-05       out     True     880020.0            True   
3       166377  2019-08-05       out     True     880020.0           False   
4       166377  2019-08-05       out    False     880022.0            True   
...        ...         ...       ...      ...          ...             ...   
53897   168606  2019-11-10       out     True     957922.0            True   
53898   168606  2019-11-11       out     True     957922.0           False   
53899   168606  2019-11-15       out     True     957922.0           False   
53900   168606  2019-11-15       out     True     957922.0           False   
53901   168606  2019-11-19        in    False          NaN            True   

       calls_count  call_duration  total_call_duration  
0                2              0                    4  
1                3              0                    5  
2                1              0                    1  
3                1             10                   18  
4                3              0                   25  
...            ...            ...                  ...  
53897            1              0                   38  
53898            2            479                  501  
53899            4           3130                 3190  
53900            4           3130                 3190  
53901            2              0                   64  

[53902 rows x 9 columns]

In [15]:
#checar valores nulos
df2.isnull().sum()

user_id                   0
date                      0
direction                 0
internal                117
operator_id            8172
is_missed_call            0
calls_count               0
call_duration             0
total_call_duration       0
dtype: int64

En la columna internal se puede ver que solo hay 117 valores nulos, concidero que es una buena opción mantener estos ya que se puede también analizar si la falta de registro lleva algún patrón. 

La columna presenta un total de 8172 valores nulos, esto lo encuentro un poco preocupante ya que el principal objetivo es valorar el desempeño de los trabajadores. 
Considero que una buena opción para este caso cambiarlos por un valor "0", más adelante podrían ser huzados para poder mostrar la cantidad de llamadas con un operador sin id, para no perder datos y mantener un volumen aunque no se pueda identificar al operador. 

In [16]:
#convertí Nan a 0 
df2["operator_id"] = df2["operator_id"].fillna("Unknown")

df2.sample(5)

user_id        date direction internal operator_id  is_missed_call  \
40133   167799  2019-10-30       out    False    925104.0            True   
14515   166837  2019-10-31        in    False    919896.0           False   
28443   167180  2019-11-07       out    False    905932.0           False   
14756   166843  2019-09-17        in    False    895172.0           False   
51899   168361  2019-10-23       out    False    945294.0            True   

       calls_count  call_duration  total_call_duration  
40133           11              0                  296  
14515            1            115                  128  
28443            7           1497                 1532  
14756            2            749                  765  
51899           66              0                 1334

In [17]:
#hacer los float a int
df2["operator_id"] = df2["operator_id"].apply(
    lambda x: int(float(x)) if str(x).replace('.', '', 1).isdigit() else "Unknown"
)

df2.sample(10)


user_id        date direction internal operator_id  is_missed_call  \
12182   166717  2019-11-08       out    False      891988           False   
11245   166691  2019-09-10        in    False     Unknown            True   
19248   166983  2019-10-25       out    False      901886           False   
19181   166983  2019-10-20        in    False      901880           False   
47417   168187  2019-10-22       out    False      937736            True   
26895   167158  2019-11-05        in    False     Unknown            True   
5265    166520  2019-09-01        in    False     Unknown            True   
47415   168187  2019-10-22        in     True     Unknown            True   
16663   166916  2019-10-25        in    False      906400           False   
20979   167011  2019-09-20       out    False      899788            True   

       calls_count  call_duration  total_call_duration  
12182            8            929                 1025  
11245            2              0                    2  
19248           24           4292                 4576  
19181           19           3120                 3212  
47417           31              0                  822  
26895            3              0                   53  
5265             2              0                   81  
47415            2              0                   31  
16663            4            297                  385  
20979            5              0                  136

## Anlisis exploratorio (EDA)

In [18]:
#Resumen estadístico 
df1.describe(include="all")

user_id tariff_plan                     date_start
count      732.000000         732                            732
unique            NaN           3                            NaN
top               NaN           C                            NaN
freq              NaN         395                            NaN
mean    167431.927596         NaN  2019-09-14 16:39:20.655737600
min     166373.000000         NaN            2019-08-01 00:00:00
25%     166900.750000         NaN            2019-08-23 00:00:00
50%     167432.000000         NaN            2019-09-16 00:00:00
75%     167973.000000         NaN            2019-10-08 00:00:00
max     168606.000000         NaN            2019-10-31 00:00:00
std        633.810383         NaN                            NaN

Resumen estadístico general df1
#
Los identificadores de usuarios están completos, con todos los IDs caen dentro de un rángo que es lógico. La mayoría de los usuarios se encuentran en el plan C. Los ususarios se distribuyen de manera continua entre agosto y octubre de 2019. 

In [19]:
df2.describe(include="all")

user_id        date direction internal operator_id  \
count    53902.000000       53902     53902    53785       53902   
unique            NaN         119         2        2        1093   
top               NaN  2019-11-25       out    False     Unknown   
freq              NaN        1220     31917    47621        8172   
mean    167295.344477         NaN       NaN      NaN         NaN   
std        598.883775         NaN       NaN      NaN         NaN   
min     166377.000000         NaN       NaN      NaN         NaN   
25%     166782.000000         NaN       NaN      NaN         NaN   
50%     167162.000000         NaN       NaN      NaN         NaN   
75%     167819.000000         NaN       NaN      NaN         NaN   
max     168606.000000         NaN       NaN      NaN         NaN   

       is_missed_call   calls_count  call_duration  total_call_duration  
count           53902  53902.000000   53902.000000         53902.000000  
unique              2           NaN            NaN                  NaN  
top             False           NaN            NaN                  NaN  
freq            30334           NaN            NaN                  NaN  
mean              NaN     16.451245     866.684427          1157.133297  
std               NaN     62.917170    3731.791202          4403.468763  
min               NaN      1.000000       0.000000             0.000000  
25%               NaN      1.000000       0.000000            47.000000  
50%               NaN      4.000000      38.000000           210.000000  
75%               NaN     12.000000     572.000000           902.000000  
max               NaN   4817.000000  144395.000000        166155.000000

Podemos ver que la mayoría de las llamadas son salientes y externas, podría ser por que principalmente son llamadas de clientes externos. Una cantidad muy baja de llamadas no tiene operator_id y aproximadamente el 44% de las llamadas fueron perdidas. Vemos una variabilidad en la cantidad de llamadaspor regirtro algo grande, hay llamadas de 0 segundos y otras extremadamente largas lo que indica que la presencia de outliers podrían mostrar un sesgo. 
